### PR TITLE
Assignment7+8-LINQ Andree and Cynthia

### PR DESCRIPTION
**Pull request by Aramirez61**
@montalcy @Aramirez61

### CODE REVIEW

**Comment by pja6** [comment]
will review

File: Assignment/SampleData.cs [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using Assignment.Interfaces;
+
+namespace Assignment;
+
+public class SampleData : ISampleData
+{
+    // 1.
```
Comment (pja6) [inline Assignment/SampleData.cs:12]: really cool checking if the file exists

File: Assignment/SampleData.cs [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+            var stateSet = new HashSet<string>();
+            foreach (var row in csvRows)
+            {
+                var columns = row.Split(',');
+                if (columns.Length >= 7) // Assuming state is at index 6
+                {
+                    stateSet.Add(columns[6].Trim()); // Trim to remove any extra spaces
+                }
+            }
+
```
Comment (pja6) [inline Assignment/SampleData.cs:49]: I'm not sure if this is using linq to accomplish the specification? this is what we did:

```suggestion
return CsvRows.Select(item => item.Split(',')[6])
       .OrderBy(state => state)
       .Distinct();//for unique

```

File: Assignment/SampleData.cs [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+            // Convert to a list and sort alphabetically
+            var sortedStates = new List<string>(stateSet);
+            sortedStates.Sort(StringComparer.OrdinalIgnoreCase); // Case-insensitive sorting
+
```
Comment (pja6) [inline Assignment/SampleData.cs:53]: Does this handle duplicates? will the output only return unique values?

File: Assignment/SampleData.cs [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+            return sortedStates;
+        }
+    // 3.
```
Comment (pja6) [inline Assignment/SampleData.cs:56]: this still passes our test without using .ToArray(), cool!

File: Assignment.Tests/SampleDataTests.cs [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+
+    [TestMethod]
+    public void FilterByEmailAddress_FilterByEducationEmail_ReturnsPeopleWithGovEmails()
+    {
+
+        Assert.IsTrue(new List<(string, string)> { ("Amelia" , "Toal"), ("Arthur" ,"Myles"), ("Ev", "Challace"), ("Marijn", "McKennan"), ("Priscilla", "Jenyns") }.SequenceEqual(SampleData.FilterByEmailAddress((email) => email.Contains(".gov")).OrderBy(fullName => fullName.FirstName)));
+
+    }
+
+
```
Comment (SaaWest) [inline Assignment.Tests/SampleDataTests.cs:93]: I believe your missing a test for GetAggregateSortedListOfStatesUsingCsvRows()

File: Assignment.Tests/SampleDataTests.cs [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
```
Comment (SaaWest) [inline Assignment.Tests/SampleDataTests.cs:1]: Some more testing conditions would be good

File: Assignment/SampleData.cs [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+    public string GetAggregateSortedListOfStatesUsingCsvRows()
+    {
+        var uniqueStates = GetUniqueSortedListOfStatesGivenCsvRows();
+        return string.Join(", ", uniqueStates);
```
Comment (SaaWest) [inline Assignment/SampleData.cs:60]: need to cast toArray()

**Review by SaaWest — COMMENTED** [review]
Implement the ISampleData.CsvRows property, loading the data from the People.csv file and returning each line as a single string. ✔
        Change the "Copy to" property on People.csv to "Copy if newer" so that the file is deployed along with your test project. ❌
        Using LINQ, skip the first row in the People.csv. ✔
        Be sure to appropriately handle resource (IDisposable) items correctly if applicable (and it may not be depending on how you implement it). ✔

    Implement IEnumerable<string> GetUniqueSortedListOfStatesGivenCsvRows() to return a sorted, unique list of states. ✔
        Use ISampleData.CsvRows for your data source. ✔
        Don't forget the list should be unique. ❌
        Sort the list alphabetically. ✔
        Include a test that leverages a hardcoded list of addresses. ❌
        Include a test that uses LINQ to verify the data is sorted correctly (do not use a hardcoded list). ✔

    Implement ISampleData.GetAggregateSortedListOfStatesUsingCsvRows() to return a string that contains a unique, comma separated list of states. ✔
        Use ISampleData.GetUniqueSortedListOfStatesGivenCsvRows() for your data source. ✔
        Consider "selecting" only the states and calling ToArray() to retrieve an array of all the state names. ❌ need toArray()
        Given the array, consider using string.Join to combine the list into a single string. ✔

    Implement the ISampleData.People property to return all the items in People.csv as Person objects ✔
        Use ISampleData.CsvRows as the source of the data. ✔
        Sort the list by State, City, and Zip. (Sort the addresses first then select). ✔
        Be sure that Person.Address is also populated. ✔
        Adding null validation to all the Person and Address properties is optional.
        Consider using ISampleData.CsvRows in your test to verify your results. ✔

    Implement ISampleDate.FilterByEmailAddress(Predicate<string> filter) to return a list of names where the email address matches the filter. ✔
        Use ISampleData.People for your data source. ✔

    Implement ISampleData.GetAggregateListOfStatesGivenPeopleCollection(IEnumerable<IPerson> people) to return a string that contains a unique, comma-separated list of states. ✔
        Use the people parameter from ISampleData.People property for your data source. ✔
        At a minimum, use the System.Linq.Enumerable.Aggregate` LINQ method to create your result. ✔
        Don't forget the list should be unique. ✔
        It is recommended that, at a minimum, you use ISampleData.GetUniqueSortedListOfStatesGivenCsvRows to validate your result.

    Given the implementation of Node in Assignment5

    Implement IEnumerable<T> to return all the items in the "circle" of items. ✔
    Add an IEnumberable<T> ChildItems(int maximum) method to Node that returns the remaining items with a maximum number of items returned less than maximum.

Extra Credit

    Implement the homework using async/await and multi-threading by defining a new SampleDataAsync class that implements IAsyncSampleData). Refactor your SampleData and SampleDataAsync classes with minimal duplication. Be sure to refactor your tests to re-use a significant amount of the test code for both implementations. ❌

Fundamentals

    Place all shared project properties into a Directory.Build.Props file.
    Place all shared project items into a Directory.Build.targets file.
    Ensure nullable reference types is enabled ✔
    Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
    Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
    and enabled .NET analyzers for both projects ✔
    For this assignment, consider using Assert.AreEqual<T>() (the generic version) ✔
    All of the above should be unit tested ❌
    Choose simplicity over complexity ✔

File: Assignment/SampleData.cs [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+    }
+
```
Comment (pja6) [inline Assignment/SampleData.cs:62]: nice job!

File: Assignment/SampleData.cs [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+    // 5.
+    public IEnumerable<(string FirstName, string LastName)> FilterByEmailAddress(Predicate<string> filter) {
+            //searching the people list for matching filter predicate, then selecting the tuple firstname, lastname and returning it
+            
+            IEnumerable<(string Fn, string Ln)> matches = People
+                .Where(currentPerson => filter(currentPerson.EmailAddress))
+                .Select(personFullName => (personFullName.FirstName, personFullName.LastName));
+            return matches;
+        }
```
Comment (pja6) [inline Assignment/SampleData.cs:87]: nice job - we used toList() in our method but i'm not sure how much of a difference that makes, checking it against the test for our method passes !

File: Assignment/SampleData.cs [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+            //using people, we select the state and trim, order by state, choosing only distinct states (making this list unique),
+            //we use aggregate to take the first persons state with the second persons state adding a comma in between. 
+            string statesList = People
+                .Select(person => person.Address.State.Trim())
+                .OrderBy(state => state)
+                .Distinct()
+                .Aggregate((person1state, person2state) => person1state + ", " + person2state);
+            return statesList;
+        }           
+
```
Comment (pja6) [inline Assignment/SampleData.cs:101]: Did basically the same thing - your use of linq feels more concise though being able to do it all at once like that - and you didn't need to change your order by statement!

File: Assignment/Node.cs [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+                if (headNode.Value?.Equals(value) ?? false) // Handling possible null reference
+                {
+                    return true;
+                }
+                headNode = headNode.Next!;
+            } while (headNode != this);
+
+            return false;
+        }
+
```
Comment (pja6) [inline Assignment/Node.cs:69]: This is how we implemented it also

**Review by pja6 — COMMENTED** [review]
Instructions

1. Implement the ISampleData.CsvRows property, loading the data from the People.csv file and returning each line as a single string. ✔
    * Change the "Copy to" property on People.csv to "Copy if newer" so that the file is deployed along with your test project. ❌✔ //wasn’t sure how to check this in GitHub
    * Using LINQ, skip the first row in the People.csv. ✔
    * Be sure to appropriately handle resource (IDisposable) items correctly if applicable (and it may not be depending on how you implement it). ✔
2. Implement IEnumerable<string> GetUniqueSortedListOfStatesGivenCsvRows() to return a sorted, unique list of states. ❌ //don’t think it handled duplicate cases
    * Use ISampleData.CsvRows for your data source. ✔
    * Don't forget the list should be unique. ❌
    * Sort the list alphabetically. ✔
    * Include a test that leverages a hardcoded list of addresses. ❌ //missing test
    * Include a test that uses LINQ to verify the data is sorted correctly (do not use a hardcoded list). ✔
3. Implement ISampleData.GetAggregateSortedListOfStatesUsingCsvRows() to return a string that contains a unique, comma separated list of states. ✔ // because this uses the previous method that might not return a unique list i’m not sure if this actually returns a unique list, it passed our test but that might just be a hole in our testing
    * Use ISampleData.GetUniqueSortedListOfStatesGivenCsvRows() for your data source. ✔
    * Consider "selecting" only the states and calling ToArray() to retrieve an array of all the state names. ❌ //toArray not used
    * Given the array, consider using string.Join to combine the list into a single string. ✔
4. Implement the ISampleData.People property to return all the items in People.csv as Person objects ✔
    * Use ISampleData.CsvRows as the source of the data. ✔
    * Sort the list by State, City, and Zip. (Sort the addresses first then select). ✔
    * Be sure that Person.Address is also populated. ✔
    * Adding null validation to all the Person and Address properties is optional.
    * Consider using ISampleData.CsvRows in your test to verify your results. ✔
5. Implement ISampleDate.FilterByEmailAddress(Predicate<string> filter) to return a list of names where the email address matches the filter. ✔
    * Use ISampleData.People for your data source. ✔
6. Implement ISampleData.GetAggregateListOfStatesGivenPeopleCollection(IEnumerable<IPerson> people) to return a string that contains a unique, comma-separated list of states. ✔
    * Use the people parameter from ISampleData.People property for your data source. ✔
    * At a minimum, use the System.Linq.Enumerable.Aggregate` LINQ method to create your result. ✔
    * Don't forget the list should be unique. ✔
    * It is recommended that, at a minimum, you use ISampleData.GetUniqueSortedListOfStatesGivenCsvRows to validate your result.
7. Given the implementation of Node in Assignment5
* Implement IEnumerable<T> to return all the items in the "circle" of items. ✔
* Add an IEnumberable<T> ChildItems(int maximum) method to Node that returns the remaining items with a maximum number of items returned less than maximum.

Fundamentals

* Place all shared project properties into a Directory.Build.Props file.✔
* Place all shared project items into a Directory.Build.targets file.
* Ensure nullable reference types is enabled ✔
* Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
* Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
* and enabled .NET analyzers for both projects ✔
* For this assignment, consider using Assert.AreEqual<T>() (the generic version) ❌✔ //unsure if MSTest handles this by default
* All of the above should be unit tested ❌ // missing some tests
* Choose simplicity over complexity ✔


Overall really solid - made a few comments, just missing a few tests and your GetUniqueSortedListOfStatesGivenCsvRows was implemented a little differently/without linq so might be something to refactor if needed

File: Assignment/SampleData.cs (hunk 3 above)
Comment (Aramirez61) [reply Assignment/SampleData.cs:53]: Implemented the change you suggested and it should return only unique values now with the distinct call.

File: Assignment/SampleData.cs (hunk 7 above)
Comment (Aramirez61) [reply Assignment/SampleData.cs:46]: Thanks for the reminder!

File: Assignment.Tests/SampleDataTests.cs (hunk 6 above)
Comment (Aramirez61) [reply Assignment.Tests/SampleDataTests.cs:1]: For sure!

File: Assignment/SampleData.cs [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+    // 4.
+    public IEnumerable<IPerson> People => CsvRows
+    .Select(line => line.Split(","))
+    .Select(columns => new
+    {
+        Line = columns,
+        Address = new Address(columns[4], columns[5], columns[6], columns[7]),
+        Person = new Person(columns[1], columns[2], null!, columns[3]) // Address will be set below
+    })
+    .OrderBy(item => (item.Line[5], item.Line[6], item.Line[7]))
```
[the remark this answers is not in the document]
Comment (Aramirez61) [reply Assignment/SampleData.cs:58]: Thanks for the suggestion! thats an interesting way to approach this. I think we'll keep our current implementation though.

File: Assignment/SampleData.cs [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+    {
+        item.Person.Address = item.Address;
+        return item.Person;
+    });
+
+    // 5.
+    public IEnumerable<(string FirstName, string LastName)> FilterByEmailAddress(Predicate<string> filter) {
+            //searching the people list for matching filter predicate, then selecting the tuple firstname, lastname and returning it
+            
+            IEnumerable<(string Fn, string Ln)> matches = People
```
[the remark this answers is not in the document]
Comment (Aramirez61) [reply Assignment/SampleData.cs:69]: Thats a good way to do that too!

File: Assignment/SampleData.cs [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+        }
+    // 3.
+    public string GetAggregateSortedListOfStatesUsingCsvRows()
+    {
+        var uniqueStates = GetUniqueSortedListOfStatesGivenCsvRows();
+        return string.Join(", ", uniqueStates);
+    }
+
+    // 4.
+    public IEnumerable<IPerson> People => CsvRows
```
[the remark this answers is not in the document]
Comment (Aramirez61) [reply Assignment/SampleData.cs]: Thats much cleaner! Just implemented that

File: Assignment/SampleData.cs (hunk 9 above)
Comment (montalcy) [reply Assignment/SampleData.cs:72]: Thanks for checking against your code as well :)

File: Assignment/SampleData.cs (hunk 10 above)
Comment (montalcy) [reply Assignment/SampleData.cs:86]: Thanks!

File: Assignment/SampleData.cs (hunk 8 above)
Comment (montalcy) [reply Assignment/SampleData.cs:48]: Thanks!

File: Assignment/SampleData.cs (hunk 4 above)
Comment (montalcy) [reply Assignment/SampleData.cs:42]: Why is it that you used toArray?

File: Assignment.Tests/SampleDataTests.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (montalcy) [reply Assignment.Tests/SampleDataTests.cs]: Looking into this now, will try and see if we can implement it, thanks for the advice!

File: Assignment.Tests/SampleDataTests.cs [hunk 15]
```diff
@@ -0,0 +1,97 @@
...
+        // Convert people objects into formatted strings representing people's data
+        IEnumerable<string> personsFromObjects = SampleData.People
+            .Select(person =>
+                $"{person.FirstName},{person.LastName},{person.EmailAddress},{person.Address.StreetAddress},{person.Address.City},{person.Address.State},{person.Address.Zip}");
+
+        // Assert that both sequences of formatted strings are equal
+        Assert.IsTrue(personsFromCsv.SequenceEqual(personsFromObjects));
+    }
+
+
```
[the remark this answers is not in the document]
Comment (montalcy) [reply Assignment.Tests/SampleDataTests.cs:66]: We will try and implement this as well, thanks for the head up!

File: Assignment.Tests/NodeTests.cs [hunk 16]
```diff
@@ -0,0 +1,149 @@
...
+            headNode.Append("second");
+            headNode.Append("third");
+            headNode.Append("fourth");
+            headNode.Clear();
+            Assert.AreEqual("start", headNode.Next.Value);
+            Assert.IsFalse(headNode.Exists("second"));
+            Assert.IsFalse(headNode.Exists("third"));
+            Assert.IsFalse(headNode.Exists("fourth"));
+        }
+
```
[the remark this answers is not in the document]
Comment (montalcy) [reply Assignment.Tests/NodeTests.cs:93]: Will be adding in this check here soon, thanks again!

File: Assignment.Tests/NodeTests.cs [hunk 17]
```diff
@@ -0,0 +1,149 @@
...
+            node3.SetNext(node1); // Create a circular linked list
+
+            // Act
+            var childItems = node1.ChildItems(maximum: 4); // Requesting more items than available
+
+            // Assert
+            var expectedItems = new int[] { 1, 2, 3 };
+            CollectionAssert.AreEqual(expectedItems, childItems.ToArray());
+        }
+
```
[the remark this answers is not in the document]
Comment (montalcy) [reply Assignment.Tests/NodeTests.cs:136]: Thank you!

File: Assignment/Node.cs (hunk 11 above)
Comment (montalcy) [reply Assignment/Node.cs:69]: Nice!

File: Assignment/Node.cs [hunk 18]
```diff
@@ -0,0 +1,99 @@
...
+            return GetEnumerator();
+        }
+
+        public IEnumerable<T> ChildItems(int maximum)
+        {
+            Node<T>? currentNode = this; // Nullable reference type
+            int count = 0;
+            do
+            {
+                if (count >= maximum)
```
[the remark this answers is not in the document]
Comment (montalcy) [reply Assignment/Node.cs:91]: I remember we covered something like this where I think he said a while loop would be preferred but I think they both make sense in this case.

File: Assignment/SampleData.cs (hunk 1 above)
Comment (montalcy) [reply Assignment/SampleData.cs:12]: Thanks!

File: Assignment/SampleData.cs [hunk 19]
```diff
@@ -0,0 +1,89 @@
...
+
+    // 4.
+    public IEnumerable<IPerson> People
+        {
+            get
+            {
+                var people = from line in CsvRows
+                             let elements = line.Split(',')
+                             let address = new Address(elements[4], elements[5], elements[6], elements[7])
+                             select new Person(elements[1], elements[2], address, elements[3]);
```
Comment (BenjaminMichaelis) [inline Assignment/SampleData.cs:57]: generally my preference is against query expressions as they don't have quite the same advanced functionality as pure linq in some complex things.

File: Assignment.Tests/SampleDataTests.cs [hunk 20]
```diff
@@ -0,0 +1,120 @@
...
+}
+
+
+
+
+
+
+
```
Comment (BenjaminMichaelis) [inline Assignment.Tests/SampleDataTests.cs:120]:
```suggestion
    }
}
```
nit: whitespace

File: Assignment.Tests/SampleDataTests.cs [hunk 21]
```diff
@@ -0,0 +1,120 @@
...
+        Assert.IsTrue(expectedList == actualList);
+    }
+
+
+    [TestMethod]
+    public void GetAggregatedListOfStatesGivenPeopleCollection__Success()
+    {
+        string states = string.Join(", ", SampleData.GetUniqueSortedListOfStatesGivenCsvRows());
+        string givenPeople = SampleData.GetAggregateListOfStatesGivenPeopleCollection(SampleData.People);
+        Assert.IsTrue(states == givenPeople);
```
Comment (BenjaminMichaelis) [inline Assignment.Tests/SampleDataTests.cs:107]:
```suggestion
        Assert.AreEqual(states, givenPeople);
```
Use Assert methods over doing it yourself

File: Assignment.Tests/SampleDataTests.cs [hunk 22]
```diff
@@ -0,0 +1,120 @@
...
+        Assert.IsTrue(new List<(string, string)> { ("Amelia" , "Toal"), ("Arthur" ,"Myles"), ("Ev", "Challace"), ("Marijn", "McKennan"), ("Priscilla", "Jenyns") }.SequenceEqual(SampleData.FilterByEmailAddress((email) => email.Contains(".gov")).OrderBy(fullName => fullName.FirstName)));
+
+    }
+
+    [TestMethod]
+    public void GetAggregateSortedListOfStatesUsingCsvRows_GivenOrderedList_Equal()
+    {
+        string expectedList = "AL, AZ, CA, DC, FL, GA, IN, KS, LA, MD, MN, MO, MT, NC, NE, NH, NV, NY, OR, PA, SC, TN, TX, UT, VA, WA, WV";
+        string actualList = string.Join(" ", SampleData.GetAggregateSortedListOfStatesUsingCsvRows());
+        Assert.IsTrue(expectedList == actualList);
```
Comment (BenjaminMichaelis) [inline Assignment.Tests/SampleDataTests.cs:98]:
```suggestion
        Assert.AreEqual(expectedList, actualList);
```
Use Assert methods over doing it yourself

File: Assignment.Tests/SampleDataTests.cs [hunk 23]
```diff
@@ -0,0 +1,120 @@
...
+
+        Assert.IsTrue(new List<(string, string)> { ("Amelia" , "Toal"), ("Arthur" ,"Myles"), ("Ev", "Challace"), ("Marijn", "McKennan"), ("Priscilla", "Jenyns") }.SequenceEqual(SampleData.FilterByEmailAddress((email) => email.Contains(".gov")).OrderBy(fullName => fullName.FirstName)));
+
+    }
+
+    [TestMethod]
+    public void GetAggregateSortedListOfStatesUsingCsvRows_GivenOrderedList_Equal()
+    {
+        string expectedList = "AL, AZ, CA, DC, FL, GA, IN, KS, LA, MD, MN, MO, MT, NC, NE, NH, NV, NY, OR, PA, SC, TN, TX, UT, VA, WA, WV";
+        string actualList = string.Join(" ", SampleData.GetAggregateSortedListOfStatesUsingCsvRows());
```
Comment (BenjaminMichaelis) [inline Assignment.Tests/SampleDataTests.cs:97]: What is this join doing if you are returning a string?

**Pull request closed without merging**